source/extensions/stat_sinks/common/statsd/statsd.cc

```diff
@@ -195,33 +258,57 @@ TcpStatsdSink::TcpStatsdSink(const LocalInfo::LocalInfo& local_info,
 void TcpStatsdSink::flush(Stats::MetricSnapshot& snapshot) {
   TlsSink& tls_sink = tls_->getTyped<TlsSink>();
   tls_sink.beginFlush(true);
-  for (const auto& counter : snapshot.counters()) {
-    if (counter.counter_.get().used()) {
-      tls_sink.flushCounter(counter.counter_.get().name(), counter.delta_);
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.enable_sanitization_during_sink")) {
+    std::string buffer;
+    for (const auto& counter : snapshot.counters()) {
+      if (counter.counter_.get().used()) {
+        tls_sink.flushCounter(sanitizeStatsdName(counter.counter_.get().name(), buffer).data(),
+                              counter.delta_);
+      }
     }
-  }
 
-  for (const auto& gauge : snapshot.gauges()) {
-    if (gauge.get().used()) {
-      tls_sink.flushGauge(gauge.get().name(), gauge.get().value());
+    for (const auto& gauge : snapshot.gauges()) {
+      if (gauge.get().used()) {
+        tls_sink.flushGauge(sanitizeStatsdName(gauge.get().name(), buffer).data(),
+                            gauge.get().value());
+      }
+    }
+  } else {
+
+    for (const auto& counter : snapshot.counters()) {
+      if (counter.counter_.get().used()) {
+        tls_sink.flushCounter(counter.counter_.get().name(), counter.delta_);
+      }
+    }
+
+    for (const auto& gauge : snapshot.gauges()) {
+      if (gauge.get().used()) {
+        tls_sink.flushGauge(gauge.get().name(), gauge.get().value());
+      }
     }
   }
   // TODO(efimki): Add support of text readouts stats.
   tls_sink.endFlush(true);
 }
 
 void TcpStatsdSink::onHistogramComplete(const Stats::Histogram& histogram, uint64_t value) {
+  std::string histogram_name;
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.enable_sanitization_during_sink")) {
```

Comment:
any reason not to capture sink_sanitization_enabled_ in the ctor like the other sinks?

Reply:
I used frequency of calls to runtimeFeatureEnabled within a single statsd sink call as criteria to latch the runtime bit to object lifetime. For example UdpStatsdSink::flush () sanitizes for each metric name by calling a separate lower level function buildMessage() and this bit is helpful within it vs in Tcp StatsdSink::flush() where single runtimeFeatureEnabled check happens before sanitization for all stats so cost of lookup is very low.

Tcp/Udp::onHistogramComplete() sink calls are done for each histogram so we may still get some benefit using ctor pattern here as bit will be set for life of the sink object. I will make the change.